op-e2e/actions/l2_validator_test.go

```diff
@@ -119,3 +125,108 @@ func RunValidatorTest(gt *testing.T, deltaTimeOffset *hexutil.Uint64) {
 }
+
+func RunValidatorManagerTest(gt *testing.T, deltaTimeOffset *hexutil.Uint64) {
+	t := NewDefaultTesting(gt)
+	dp := e2eutils.MakeDeployParams(t, defaultRollupTestParams)
+	dp.DeployConfig.L2GenesisDeltaTimeOffset = deltaTimeOffset
+	sd := e2eutils.Setup(t, dp, defaultAlloc)
+	log := testlog.Logger(t, log.LvlDebug)
+	miner, seqEngine, sequencer := setupSequencerTest(t, sd, log)
+
+	rollupSeqCl := sequencer.RollupClient()
+	batcher := NewL2Batcher(log, sd.RollupCfg, DefaultBatcherCfg(dp),
+		rollupSeqCl, miner.EthClient(), seqEngine.EthClient(), seqEngine.EngineClient(t, sd.RollupCfg))
+
+	validator := NewL2Validator(t, log, &ValidatorCfg{
+		OutputOracleAddr:     sd.DeploymentsL1.L2OutputOracleProxy,
+		ValidatorPoolAddr:    sd.DeploymentsL1.ValidatorPoolProxy,
+		ValidatorManagerAddr: sd.DeploymentsL1.ValidatorManagerProxy,
+		AssetManagerAddr:     sd.DeploymentsL1.AssetManagerProxy,
+		ColosseumAddr:        sd.DeploymentsL1.ColosseumProxy,
+		SecurityCouncilAddr:  sd.DeploymentsL1.SecurityCouncilProxy,
+		ValidatorKey:         dp.Secrets.TrustedValidator,
+		AllowNonFinalized:    false,
+	}, miner.EthClient(), seqEngine.EthClient(), sequencer.RollupClient())
+
+	for i := 0; i < 5; i++ {
+		// L1 block
+		miner.ActEmptyBlock(t)
+		// L2 block
+		sequencer.ActL1HeadSignal(t)
+		sequencer.ActL2PipelineFull(t)
+		sequencer.ActBuildToL1Head(t)
+		// submit and include in L1
+		batcher.ActSubmitAll(t)
+		miner.includeL1Block(t, dp.Addresses.Batcher, 12)
+		// finalize the first and second L1 blocks, including the batch
+		miner.ActL1SafeNext(t)
+		miner.ActL1SafeNext(t)
+		miner.ActL1FinalizeNext(t)
+		miner.ActL1FinalizeNext(t)
+		// derive and see the L2 chain fully finalize
+		sequencer.ActL2PipelineFull(t)
+		sequencer.ActL1SafeSignal(t)
+		sequencer.ActL1FinalizedSignal(t)
+	}
+
+	// deposit bond for validator
+	validator.ActDeposit(t, 1_000)
+	miner.includeL1Block(t, validator.address, 12)
+	// Submit 16 outputs to ValidatorPool
+	for i := 0; i < 16; i++ {
+		// submit to L1
+		validator.ActSubmitL2Output(t)
+		// include output on L1
+		miner.includeL1Block(t, validator.address, 12)
+		miner.ActEmptyBlock(t)
+		// Check submission was successful
+		receipt, err := miner.EthClient().TransactionReceipt(t.Ctx(), validator.LastSubmitL2OutputTx())
+		require.NoError(t, err)
+		require.Equal(t, types.ReceiptStatusSuccessful, receipt.Status, "submission failed")
+	}
+
+	// approve governance token
+	validator.ActApprove(t, 1_000)
+	miner.includeL1Block(t, validator.address, 12)
+
+	// register validator
+	validator.ActRegisterValidator(t, new(big.Int).SetUint64(1_000))
+	miner.includeL1Block(t, validator.address, 12)
+
+	require.Equal(t, sequencer.SyncStatus().UnsafeL2, sequencer.SyncStatus().FinalizedL2)
+	// create l2 output submission transactions until there is nothing left to submit
+	for {
+		waitTime := validator.CalculateWaitTime(t)
+		if waitTime > 0 {
+			break
+		}
+		// and submit it to L1
+		validator.ActSubmitL2Output(t)
```

Comment:
I think this code block is not reachable. Maybe need to proceed more blocks after submitting outputs to ValidatorPool.